### PR TITLE
library: Adds File Monitor Entry

### LINKED ISSUE #495: File Monitor
https://docs.gtk.org/gio/method.File.monitor.html

We could have a simple UI that invites the user to save a `monitored.txt` file, invite the user to edit it and display something when the file is edited.

Not sure what we can do about folder.

### PR DESCRIPTION
**Pull request by SoNiC-HeRE**
Closes #495

### CODE REVIEW

**Comment by SoNiC-HeRE** [comment]
[Kooha-2023-08-18-00-56-10.webm](https://github.com/sonnyp/Workbench/assets/96797205/9062a09f-5330-44a5-b51e-f42f187d7f08)

**Comment by andyholmes** [comment]
I couldn't get this to work, it seems as though there are no permissions for document. Did you have to do anything special to get this to work?

**Comment by SoNiC-HeRE** [comment]
> I couldn't get this to work, it seems as though there are no permissions for document. Did you have to do anything special to get this to work?

Nope, Although i had to run the demo quite a no of times to get this demo working, not sure why is it happening though

**Comment by andyholmes** [comment]
Okay, so I figured out what's happening here:

image

Notice in your video the path begins with `~/.var/`, which means that when you recorded the screencast, `workbench.txt` was in a user folder. When the demo is first loaded however, it's in the Flatpak's `/app` folder:

```sh
$ ls -la /app/share/re.sonny.Workbench.Devel/Library/demos/File\ Monitor/
total 20
drwxr-xr-x. 1 andrew andrew   88 Aug 19 08:39 .
drwxr-xr-x. 1 andrew andrew 1804 Aug 19 08:39 ..
-rw-r--r--. 1 andrew andrew  699 Aug 19 04:29 main.blp
-rw-r--r--. 1 andrew andrew  952 Aug 19 04:29 main.js
-rw-r--r--. 1 andrew andrew  192 Aug 19 04:29 main.json
-rw-r--r--. 1 andrew andrew 1922 Aug 19 08:39 main.ui
-rw-r--r--. 1 andrew andrew   48 Aug 19 04:29 workbench.txt
```

So here's what I think the best thing to do is:

1. Use [`Gio.File.new_tmp_dir_async()`](https://gjs-docs.gnome.org/gio20/gio.file#function-new_tmp_dir_async)
2. Create a `Gio.FileMonitor` for the returned `Gio.File`, and connect to the `changed` singal with `Gio.FileMonitorFlags` set to watch for new files.
3. Add a button that says "Create File", which calls [`Gio.File.create_async()`](https://gjs-docs.gnome.org/gio20~2.0/gio.file#method-create_async) on a child path (e.g. `file.get_child('workbench.txt').create_async(...)`, using a `.catch(logError)` instead of `await`
5. In the callback of the above `changed` signal, take the `other_file` argument. Have a second button that opens the file launcher for that file.

Does that all make sense?

**Comment by andyholmes** [comment]
> That shouldn't be the case, demos are always copied into `~/.var` IIRC

If that's so, I can't explain why the demo won't work :confused: 

[Screencast from 2023-08-19 09-16-18.webm](https://github.com/sonnyp/Workbench/assets/1265208/9061110f-4014-4b73-9ca0-6623b1c64c7c)

**Comment by SoNiC-HeRE** [comment]
Implemented:
- New UI
- 3 buttons namely: Create,Delete and View File and 1 for apply changes
- TextView for taking input from user

Issue:
- The new created file gets created in the home path and not in the current directory
- a bunch of new warnings

Also for the new file created how should i remove the previous ```file_monitor to create a new one``` or just reassign it to the new file?

File: src/Library/demos/File Monitor/main.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+const { buffer } = edit_entry;
+const file = Gio.File.new_for_path(pkg.pkgdatadir).resolve_relative_path(
+  "Library/demos/File Monitor/workbench.txt",
+);
+const file_dir = file.get_parent();
+const file_launcher = new Gtk.FileLauncher({
+  always_ask: true,
+  file,
+});
+const monitor_for_dir = file.monitor(Gio.FileMonitorFlags.NONE, null);
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:23]:
```suggestion
const monitor_for_dir = file_dir.monitor(Gio.FileMonitorFlags.WATCH_MOVES, null);
```

Don't forget to set the correct flags for what you want to monitor.

File: src/Library/demos/File Monitor/main.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+const edit_entry = workbench.builder.get_object("edit_entry");
+const view_file = workbench.builder.get_object("view_file");
+const create_file = workbench.builder.get_object("create_file");
+const delete_file = workbench.builder.get_object("delete_file");
+const edit_file = workbench.builder.get_object("edit_file");
+const file_name = workbench.builder.get_object("file_name");
+const { buffer } = edit_entry;
+const file = Gio.File.new_for_path(pkg.pkgdatadir).resolve_relative_path(
+  "Library/demos/File Monitor/workbench.txt",
+);
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:17]:
```suggestion
const file = Gio.File.new_for_uri(workbench.resolve("workbench.txt"));
```

We can use the session directory, I think.

File: src/Library/demos/File Monitor/main.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  Gio.FileQueryInfoFlags.NONE,
+  null,
+);
+file_name.label = details.get_display_name();
+buffer.set_text("Start editing ... ", -1);
+delete_file.connect("clicked", () => {
+  file.delete_async(GLib.PRIORITY_DEFAULT, null).catch(logError);
+});
+create_file.connect("clicked", () => {
+  const new_file = Gio.File.new_for_path("test-file.txt");
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:37]:
```suggestion
  const new_file = file_dir.get_child("new-file.txt");
```

File: src/Library/demos/File Monitor/main.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  new_file.create(Gio.FileCreateFlags.NONE, null);
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:38]: Make this one async as well. We can do all the reacting to the `::changed` signal.

File: src/Library/demos/File Monitor/main.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  });
+  overlay.add_toast(toast);
+});
+monitor_for_dir.connect("changed", () => {
+  const toast = new Adw.Toast({
+    title: "Directory Modified",
+    timeout: 2,
+  });
+  overlay.add_toast(toast);
+});
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:56]: This is the important part of the monitoring, determining what changed and how.

```suggestion
monitor_for_dir.connect("changed", (monitor, child, other_file, event) => {
  const toast = new Adw.Toast({
    title: "Unknown Event",
    timeout: 2,
  });

  switch (event) {
    case Gio.FileMonitorEvent.RENAMED:
      toast.title = `${child.get_basename()} was renamed to ${other_file.get_basename()}`;
      break;

    case Gio.FileMonitorEvent.MOVED_IN:
      toast.title = `${child.get_basename()} was moved into the directory`;
      break;

    case Gio.FileMonitorEvent.MOVED_OUT:
      toast.title = `${child.get_basename()} was moved out of the directory`;
      break;
  }

  overlay.add_toast(toast);
});
```

File: src/Library/demos/File Monitor/main.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  );
+  const bytes = new GLib.Bytes(text);
+  file
+    .replace_contents_async(
+      bytes,
+      null,
+      false,
+      Gio.FileCreateFlags.REPLACE_DESTINATION,
+      null,
+      null,
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:72]:
```suggestion
```

Remove to use the promisified version.

File: src/Library/demos/File Monitor/main.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+edit_file.connect("clicked", () => {
+  const text = buffer.get_text(
+    buffer.get_start_iter(),
+    buffer.get_end_iter(),
+    true,
+  );
+  const bytes = new GLib.Bytes(text);
+  file
+    .replace_contents_async(
+      bytes,
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:67]:
```suggestion
  const bytes = new GLib.Bytes(text);
  file
    .replace_contents_async(
      new GLib.Bytes(buffer.text),
```

Remember to use properties when possible. Don't forget the other methods in the file :slightly_smiling_face:

File: src/Library/demos/File Monitor/main.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
+import Gtk from "gi://Gtk";
+import Gio from "gi://Gio";
+import Adw from "gi://Adw";
+import GLib from "gi://GLib";
+
+Gio._promisify(Gtk.FileLauncher.prototype, "launch", "launch_finish");
+
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:7]: I guess we should add the other `Gio.File` operations? They're probably already done though, so I'm not sure.

File: src/Library/demos/File Monitor/main.blp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      }
+      TextView edit_entry {
+        height-request: 96;
+        width-request: 96;
+        right-margin: 12;
+        left-margin: 12;
+        top-margin: 12;
+        bottom-margin: 12;
+        wrap-mode: char;
+      }
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.blp:26]:
```suggestion
      Frame {
        height-request: 96;
        width-request: 96;
        ScrolledWindow {
          TextView edit_entry {
            right-margin: 12;
            left-margin: 12;
            top-margin: 12;
            bottom-margin: 12;
            wrap-mode: char;
          }
        }
      }
```

File: src/Library/demos/File Monitor/main.blp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    title: _("File Monitor");
+    description: _("Monitors a file or directory for changes");
+    icon-name: "paper-symbolic";
+    Box {
+      orientation: vertical;
+      halign: center;
+      spacing: 18;
+
+      Label file_name {
+        label: _("");
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.blp:15]:
```suggestion
        label: _("");
```

This is a reserved string in gettext, you can't ever use `_("")`. I guess you can leave this `null`, or set it to `_("No File")`.

**Review by andyholmes — CHANGES_REQUESTED** [review]
Good progress!

There's a little more to do, like the monitor for the file and some style cleanups, then it should be good.

File: src/Library/demos/File Monitor/main.js (hunk 8 above)
Comment (SoNiC-HeRE) [reply src/Library/demos/File Monitor/main.js:7]: I think most of them have been covered, wouldn't be worth adding them. We can add it back if that's needed

**Review by andyholmes — CHANGES_REQUESTED** [review]

File: src/Library/demos/File Monitor/main.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  file_launcher.launch(workbench.window, null).catch(logError);
+});
+
+monitor_for_file.connect("changed", () => {
+  const toast = new Adw.Toast({
+    title: "File modified",
+    timeout: 2,
+  });
+  overlay.add_toast(toast);
+});
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:57]:
```suggestion
monitor_for_file.connect("changed", (monitor, file, other_file, event) => {
  console.log(event);
  if (event === Gio.FileMonitorEvent.CHANGES_DONE_HINT) {
    const toast = new Adw.Toast({
      title: "File modified",
      timeout: 2,
    });
    overlay.add_toast(toast);
  }
});
```

File: src/Library/demos/File Monitor/main.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      toast.title = `${child.get_basename()} was renamed to ${other_file.get_basename()}`;
+      break;
+    case Gio.FileMonitorEvent.MOVED_IN:
+      toast.title = `${child.get_basename()} was moved into the directory`;
+      break;
+    case Gio.FileMonitorEvent.MOVED_OUT:
+      toast.title = `${child.get_basename()} was moved out of the directory`;
+      break;
+  }
+  overlay.add_toast(toast);
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:75]:
```suggestion
  const toast = new Adw.Toast({
    timeout: 2,
  });
  
  switch (event) {
    case Gio.FileMonitorEvent.RENAMED:
      toast.title = `${child.get_basename()} was renamed to ${other_file.get_basename()}`;
      break;
    case Gio.FileMonitorEvent.MOVED_IN:
      toast.title = `${child.get_basename()} was moved into the directory`;
      break;
    case Gio.FileMonitorEvent.MOVED_OUT:
      toast.title = `${child.get_basename()} was moved out of the directory`;
      break;
  }
  
  if (toast.title)
    overlay.add_toast(toast);
```

This is firing with a lot of "Unknown Event" toasts. On second thought, I think its better to skip `Gio.FileCreateFlags.WATCH_MOVED` and just watch regular events, so its not too complicated.

Pick the events you want to watch for the directory from the [documentation](https://gjs-docs.gnome.org/gio20/gio.filemonitorevent). I'd probably pick `DELETED` and `CREATED`.

File: src/Library/demos/File Monitor/main.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  file.delete_async(GLib.PRIORITY_DEFAULT, null).catch(logError);
+});
+
+create_file.connect("clicked", () => {
+  const new_file = file_dir.get_child("new-file.txt");
+  new_file
+    .create_async(Gio.FileCreateFlags.NONE, GLib.PRIORITY_DEFAULT, null, null)
+    .catch(logError);
+});
+
```
Comment (andyholmes) [inline src/Library/demos/File Monitor/main.js:46]:
```suggestion
```

I think I made this demo a little too complex. I'd suggest we remove this and the "create" button, and the "Apply changes" button should recreate it if necessary.

**Review by andyholmes — CHANGES_REQUESTED** [review]
I changed my mind, after thinking I made this too complicated :slightly_smiling_face:

**Review by andyholmes — APPROVED** [review]
LGTM!

**Pull request merged**